fs-storage/src/folder_storage.rs

```diff
@@ -0,0 +1,573 @@
+use serde::{Deserialize, Serialize};
+use std::fs::{self, File};
+use std::io::{Read, Write};
+use std::time::SystemTime;
+use std::{
+    collections::BTreeMap,
+    path::{Path, PathBuf},
+};
+
+use crate::base_storage::{BaseStorage, SyncStatus};
+use crate::monoid::Monoid;
+// use crate::utils::read_version_2_fs;
+use data_error::{ArklibError, Result};
+
+/*
+Note on `FolderStorage` Versioning:
+
+`FolderStorage` is a basic key-value storage system that persists data to disk.
+where the key is the path of the file inside the directory.
+
+
+In version 2, `FolderStorage` stored data in a plaintext format.
+Starting from version 3, data is stored in JSON format.
+
+For backward compatibility, we provide a helper function `read_version_2_fs` to read version 2 format.
+*/
+const STORAGE_VERSION: i32 = 3;
+
+/// Represents a folder storage system that persists data to disk.
+pub struct FolderStorage<K, V>
+where
+    K: Ord,
+{
+    /// Label for logging
+    label: String,
+    /// Path to the underlying file where data is persisted
+    path: PathBuf,
+    /// `ram_timestamps` can be used to track the last time a file was modified in memory.
+    /// where the key is the path of the file inside the directory.
+    ram_timestamps: BTreeMap<K, SystemTime>,
```

Comment:
In `FileStorage`, we used two concepts:
1. `modified`: The last time the internal mapping was changed.
2. `written_to_disk`: The last time data was saved to disk.

Do we need to define `ram_timestamps`? Is there a difference between `ram_timestamps` and the internal mapping timestamp in `FileStorage`? If not, we can stick with the existing convention from `FileStorage`.

Adding new terms might be confusing.